src/lib/http-observer.js

```diff
@@ -0,0 +1,223 @@
+import Uri from 'urijs';
+
+import Logger from './log';
+import { CDN_HEADERS, CDN_HOSTNAMES } from './cdn';
+
+const logger = new Logger('webext.lib.http-observer', console);
+
+export const REQUEST_TYPES = [
+  'main_frame',
+  'sub_frame',
+  'stylesheet',
+  'script',
+  'image',
+  'object',
+  'xmlhttprequest',
+  'xbl',
+  'xslt',
+  'ping',
+  'beacon',
+  'xml_dtd',
+  'font',
+  'media',
+  'websocket',
+  'csp_report',
+  'imageset',
+  'web_manifest',
+  'other'
+];
+
+export default class HttpObserver {
+  constructor() {
+    this.requests = {};
+    this.addListeners();
+  }
+
+  addListeners() {
+    browser.webNavigation.onBeforeNavigate.addListener(details => {
+      this.navigationListener(details);
+    });
+    browser.webRequest.onBeforeRequest.addListener(
+      details => this.requestListener(details),
+      { urls: [ '<all_urls>' ] }
+    );
+    browser.webRequest.onCompleted.addListener(
+      details => this.responseListener(details),
+      { urls: [ '<all_urls>' ] },
+      [ 'responseHeaders' ]
+    );
+  }
+
+  // Whenever the user navigates to a new page, clear the collected requests
+  // for this tab.
+  navigationListener(details) {
+    this.requests[details.tabId] = {};
+  }
+
+  // Whenever a request is made, log some initial metadata. Only do this if the
+  // instance was listening at webNavigation.onBeforeNavigate.
+  requestListener(details) {
+    if (details.tabId in this.requests) {
+      this.requests[details.tabId][details.requestId] = {
+        start: details.timeStamp,
+        url: details.url,
+        completed: false
+      };
+    }
+  }
+
+  // Whenever a request is completed, log additional metadata. Only do this if
+  // the instance was listening at webNavigation.onBeforeNavigate.
+  responseListener(details) {
+    if (details.tabId in this.requests) {
+      this.requests[details.tabId][details.requestId] = Object.assign(
+        this.requests[details.tabId][details.requestId],
```

Comment:
I thought about this, but decided not to do this because of the way JS does assignment:

```
> a = { b: 'c' };
> d = a['b']
> d = 'e'
> a
{ b: 'c' }
```

We wouldn't really save many bytes.